server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -107,15 +109,15 @@ private static void fixRedundantIncludes(ObjectMapper.Builder omb, boolean paren
                     boolean includeInRootViaParent = parentIncluded && isNested && nested.isIncludeInParent();
                     boolean includedInRoot = isNested && nested.isIncludeInRoot();
                     if (includeInRootViaParent && includedInRoot) {
-                        child.nested = Nested.newNested(true, false);
+                        child.nested = Nested.newNested(new Explicit<>(true, true), new Explicit<>(false, true));
```

Comment:
The fact that we fix redundant includes makes merging these options more tricky. Previously, mapping merges couldn't make any changes to the `include_in_root` or `include_in_parent` values. Now they can overwrite existing values, so we could have the following situation:
* A new mapper changes `include_in_root` and `include_in_parent` to `true` for some nested object.
* After merging it into the existing one, say the merged mapper now has redundant includes.
* We never detect and fix the redundancy because this method is only called in `build`, not when merging mappers.

I'm wondering if there's a lightweight way to fix this, or some way we can simplify?

Reply:
Slightly related but we were reluctant to deprecate/remove these options because Kibana doesn't support nested fields. This is something that was stalled for some time but I think we should revisit #12461 now that [nested fields are supported](https://github.com/elastic/kibana/issues/1084#issuecomment-585178079) in Kibana. I'll comment on the issue but I wanted to mention here since it changes the perspective a bit.

> I'm wondering if there's a lightweight way to fix this, or some way we can simplify?

Is there a reason not to call `fixRedundantIncludes` after merging mappers ? That seems like the easiest solution ?

Reply:
So `fixRedundantIncludes` currently works on mapper builders, not mappers themselves. I'll see if I can change this and maybe only call the method once post-mapping merge ?